zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.stream.api;
+
+import io.camunda.zeebe.stream.impl.ControllableStreamClockImpl;
+import io.camunda.zeebe.stream.impl.UncontrolledStreamClock;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.InstantSource;
+
+/**
+ * A clock that provides the current time. Used as a marker interface instead of using {@link
+ * InstantSource} directly to indicate that the clock is used for stream processing.
+ */
+public interface StreamClock extends InstantSource {
+
+  static StreamClock uncontrolled(final InstantSource source) {
+    return new UncontrolledStreamClock(source);
+  }
+
+  static StreamClock system() {
+    return new UncontrolledStreamClock(InstantSource.system());
+  }
+
+  static ControllableStreamClock controllable(final InstantSource source) {
+    return new ControllableStreamClockImpl(source);
+  }
+
+  /**
+   * A controllable {@link StreamClock} that allows to pin the time to a specific instant or to
+   * offset the current system time by some duration.
+   */
+  interface ControllableStreamClock extends StreamClock {
+
+    /**
+     * Modifies the clock by applying the given modification. Previous modifications are overridden
+     * and do not stack up.
+     *
+     * @param modification the modification to apply. Use {@link Modification#none()} to reset the
+     *     clock to current system time.
+     */
+    void applyModification(Modification modification);
```

Comment:
:question: This implies that if I say, `offset(5)`, then `offset(5)`, nothing will change? In the previous controlled clock this would actually keep incrementing by 5, as it was always a relative offset. Is this the behavior we want?

Reply:
Yes, that's also part of the docstring:
```
/**
 * Modifies the clock by applying the given modification. Previous modifications are overridden
 * and do not stack up.
 *
```

> Is this the behavior we want

I don't really know yet. If not, the desired behavior can be emulated easily with first querying current offset, adding to it and then applying a new one.

Reply:
I've added a convenience method to do that, `stackOffset`.